### PR TITLE
parquet-tools 1.8.1 (new formula)

### PR DESCRIPTION
**Pull request by kadwanev**
It's nice to have parquet-tools easily installed similar to avro-tools.

### CODE REVIEW

**Comment by DomT4** [comment]
Test is failing:
```
==> parquet-tools
No command specified

Error: parquet-tools: failed
Failed executing: parquet-tools 
```

File: Formula/parquet-tools.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  bottle :unneeded
+
+  def install
+    libexec.install "parquet-tools-#{version}.jar"
+    bin.write_jar_script libexec/"parquet-tools-#{version}.jar", "parquet-tools"
+  end
+
+  test do
+    shell_output "parquet-tools cat -h"
```
Comment (mikemcquaid) [inline Formula/parquet-tools.rb]: `system "#{bin}/parquet-tools", "cat", "-h"`

Comment (mikemcquaid) [inline Formula/parquet-tools.rb]: Also, can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

Comment (kadwanev) [inline Formula/parquet-tools.rb]: Sorry. I'm just starting to use these tools myself.

**Comment by heuermh** [comment]
Note attempting to run the jar on Maven Central linked above will fail due to missing Hadoop dependencies:
```bash
$ java -jar parquet-tools-1.8.1.jar head --debug foo.gz.parquet 
java.lang.NoClassDefFoundError: org/apache/hadoop/fs/PathFilter
	at org.apache.parquet.tools.command.HeadCommand.execute(HeadCommand.java:81)
	at org.apache.parquet.tools.Main.main(Main.java:222)
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.fs.PathFilter
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 2 more
org/apache/hadoop/fs/PathFilter
```

Per the [parquet-tools readme](https://github.com/apache/parquet-mr/blob/master/parquet-tools/README.md) the homebrew build will need to build from source with local mode, e.g.
```bash
cd parquet-tools && mvn clean package -Plocal 
```

File: Formula/parquet-tools.rb (hunk 1 above)
Comment (mikemcquaid) [inline Formula/parquet-tools.rb]: If nothing else it should use `system`. I'd rather we had a test that did something slightly more involved, too.

**Comment by kadwanev** [comment]
@heuermh 
Thanks. I changed the formula to build from source. 

I think the best thing to do for enhanced testing is to get a sample file into the repo. Alternatively, if the tools gains the ability to create a parquet file, then a full round trip can be tested.

**Comment by apjanke** [comment]
Build and audit both failed.

Build failure:

```
==> mvn clean package -Plocal
[INFO] Scanning for projects...
Downloading: https://repo.maven.apache.org/maven2/org/apache/apache/16/apache-16.pom

[WARNING] Failed to create parent directories for tracking file /Users/brew/.m2/repository/org/apache/apache/16/apache-16.pom.lastUpdated
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.apache.parquet:parquet:1.8.1: Could not transfer artifact org.apache:apache:pom:16 from/to central (https://repo.maven.apache.org/maven2): /Users/brew/.m2/repository/org/apache/apache/16/apache-16.pom.part.lock (No such file or directory) and 'parent.relativePath' points at wrong local POM @ org.apache.parquet:parquet:1.8.1, /private/tmp/parquet-tools-20160418-1258-1gkinr2/pom.xml, line 4, column 11
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.apache.parquet:parquet-tools:[unknown-version] (/private/tmp/parquet-tools-20160418-1258-1gkinr2/parquet-tools/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.apache.parquet:parquet:1.8.1: Could not transfer artifact org.apache:apache:pom:16 from/to central (https://repo.maven.apache.org/maven2): /Users/brew/.m2/repository/org/apache/apache/16/apache-16.pom.part.lock (No such file or directory) and 'parent.relativePath' points at wrong local POM @ org.apache.parquet:parquet:1.8.1, /private/tmp/parquet-tools-20160418-1258-1gkinr2/pom.xml, line 4, column 11 -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
==> Formula
Path: /usr/local/Library/Taps/homebrew/homebrew-core/Formula/parquet-tools.rb
==> Configuration
```

Perhaps an interaction between maven fetching and `superenv`?

Audit failure:

```
failed: brew audit parquet-tools --strict --online

Stacktrace
        Error: 3 problems in 1 formula

==> audit problems
parquet-tools:
 * `bottle modifier` (line 15) should be put before `depends_on` (line 5)
 * Stable: version 1.8.1 is redundant with version scanned from URL
 * Stable: Git should specify :revision when a :tag is specified.  
```

**Comment by DomT4** [comment]
> Perhaps an interaction between maven fetching and superenv?

Sandbox. Add `ENV.java_cache` to the top of `def install` and `test do` and that'll resolve that issue.

File: Formula/parquet-tools.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class ParquetTools < Formula
+  desc "ApacheParquet command-line tools and utilities"
+  homepage "https://parquet.apache.org/"
+  env :std
+
+  stable do
```
Comment (DomT4) [inline Formula/parquet-tools.rb]: You don't need either the `stable` or `head` blocks here. Can just be:

```ruby
desc "ApacheParquet command-line tools and utilities"
homepage "https://parquet.apache.org/"
url "https://github.com/apache/parquet-mr.git",
    :tag => "apache-parquet-1.8.1",
    :revision => "xyz"

head "https://github.com/apache/parquet-mr.git"
```

File: Formula/parquet-tools.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class ParquetTools < Formula
+  desc "ApacheParquet command-line tools and utilities"
+  homepage "https://parquet.apache.org/"
+  env :std
```
Comment (DomT4) [inline Formula/parquet-tools.rb]: Don't need this, and we wouldn't accept a new formula that needs `env :std`. The problem you're trying to solve I suspect can be resolved by adding `ENV.java_cache` at the top of `def install`.

**Comment by kadwanev** [comment]
Oops, looked up a "env :std" on SO and committed that. Should I change to the java_cache change right away?

File: Formula/parquet-tools.rb (hunk 3 above)
Comment (kadwanev) [inline Formula/parquet-tools.rb]: Thanks. Changing now

File: Formula/parquet-tools.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  stable do
+    url "https://github.com/apache/parquet-mr.git", :tag => "apache-parquet-1.8.1", :revision => "4aba4dae7bb0d4edbcf7923ae1339f28fd3f7fcf"
+  end
+
+  head do
+    url "https://github.com/apache/parquet-mr.git"
+  end
+
+  bottle :unneeded
```
Comment (DomT4) [inline Formula/parquet-tools.rb]: You'll need a bottle here. The `mvn` step == compile.

Comment (kadwanev) [inline Formula/parquet-tools.rb]: Instructions for that?

Comment (DomT4) [inline Formula/parquet-tools.rb]: Just remove this line, essentially :)

File: Formula/parquet-tools.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    ENV.java_cache
+    cd "parquet-tools" do
+      system "mvn", "clean", "package", "-Plocal"
+      libexec.install "target/parquet-tools-#{version}.jar"
+      bin.write_jar_script libexec/"parquet-tools-#{version}.jar", "parquet-tools"
+    end
+  end
+
+  test do
+    system "parquet-tools", "cat", "-h"
```
Comment (mikemcquaid) [inline Formula/parquet-tools.rb]: This should read a Parquet file. You can write one out in the test.

File: Formula/parquet-tools.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+class ParquetTools < Formula
+  desc "ApacheParquet command-line tools and utilities"
+  homepage "https://parquet.apache.org/"
+  url "https://github.com/apache/parquet-mr.git", :tag => "apache-parquet-1.8.1", :revision => "4aba4dae7bb0d4edbcf7923ae1339f28fd3f7fcf"
```
Comment (DomT4) [inline Formula/parquet-tools.rb]: Can you throw this onto three lines so it doesn't stretch the formula out unnecessarily? Thanks!

```ruby
url "https://github.com/apache/parquet-mr.git",
    :tag => "apache-parquet-1.8.1",
    :revision => "4aba4dae7bb0d4edbcf7923ae1339f28fd3f7fcf"
```

File: Formula/parquet-tools.rb (hunk 5 above)
Comment (DomT4) [inline Formula/parquet-tools.rb]: Please also call the test executable with a full path, i.e.:

```ruby
system bin/"parquet-tools", "cat", "-h"
```
Or:
```ruby
system "#{bin}/parquet-tools", "cat", "-h"
```
Using the current test as an example.

File: Formula/parquet-tools.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class ParquetTools < Formula
+  desc "ApacheParquet command-line tools and utilities"
```
Comment (apjanke) [inline Formula/parquet-tools.rb]: Minor nitpick: should be spelled "Apache Parquet" instead of "ApacheParquet", since that's how the project styles it.

**Comment by apjanke** [comment]
Looks good to me now, aside from the name spelling. Any other concerns?

**Comment by DomT4** [comment]
LGTM as well. When you fix the description per Andrew's request @kadwanev we'd be appreciative if you could squash back down to one commit as well. Saves us having to do that manually on pull.

There's some documentation on squashing [here](https://help.github.com/articles/about-git-rebase/) if that's helpful at all.

**Comment by apjanke** [comment]
Thanks! This is good to go now; merging.

FYI: when you make additional changes on a formula or squash commits or whatever, you can also leave a comment that you did so, for quicker response. We get notified when there are new comments on PRs we've interacted with, but not (I think) when new code changes are pushed. (E.g. I just now noticed this was ready to go when I was doing a pass through checking all the PRs in the `homebrew-core` queue.)

Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by heuermh** [comment]
Yes, thank you, @kadwanev!

**Pull request closed without merging**